packages/grafana-ui/src/components/uPlot/context.ts

```diff
@@ -0,0 +1,147 @@
+import React, { useCallback, useContext } from 'react';
+import uPlot from 'uplot';
+import { PlotPlugin } from './types';
+import { DataFrame, Field, FieldConfig } from '@grafana/data';
+
+interface PlotCanvasContextType {
+  // canvas size css pxs
+  width: number;
+  height: number;
+  // plotting area bbox, css pxs
+  plot: {
+    width: number;
+    height: number;
+    top: number;
+    left: number;
+  };
+}
+
+interface PlotContextType {
+  u: uPlot;
+  series: uPlot.Series[];
+  canvas: PlotCanvasContextType;
+}
+
+interface PlotDataContextType {
+  data: DataFrame;
+}
+
+type PlotPluginsContextType = {
+  registerPlugin: (plugin: PlotPlugin) => () => void;
+};
+
+export const PlotContext = React.createContext<PlotContextType | null>(null);
+export const PlotPluginsContext = React.createContext<PlotPluginsContextType>(null);
+export const PlotDataContext = React.createContext<PlotDataContextType | null>(null);
+
+// Exposes uPlot instance and bounding box of the entire canvas and plot area
+export const usePlotContext = (): PlotContextType | null => {
+  return useContext<PlotContextType | null>(PlotContext);
+};
+
+// Exposes API for registering uPlot plugins
+export const usePlotPluginContext = (): PlotPluginsContextType => {
+  return useContext(PlotPluginsContext);
+};
+
+interface PlotDataAPI {
+  /** Data frame passed to graph, x-axis aligned */
+  data: DataFrame;
+  /** Returns field by index */
+  getField: (idx: number) => Field;
+  /** Returns x-axis fields */
+  getXAxisFields: () => Field[];
+  /** Returns x-axis fields */
+  getYAxisFields: () => Field[];
+  /** Returns field value by field and value index */
+  getFieldValue: (fieldIdx: number, rowIdx: number) => any;
+  /** Returns field config by field index */
+  getFieldConfig: (fieldIdx: number) => FieldConfig;
```

Comment:
I am confused by this API -- are the fields each thing returns different than what shows up in `data`?  

can `getField(2)` be different than `data.fields[2]`?

Should the X/Y getters return indexes?  or should we not have data and assert that everything is the same length?